Add saved-query API

xerial opened the pull request.

danielnorberg commented on src/main/java/com/treasuredata/client/model/TDSavedQueryBuilder.java:
```diff
@@ -0,0 +1,159 @@
+                result.or(base.getResult()));
+    }
+
+    public static <T> void checkPresence(Optional<T> opt, String errorMessage)
```
Should this method be private?

danielnorberg commented on src/main/java/com/treasuredata/client/TDClient.java:
```diff
@@ -627,14 +637,78 @@ public void deleteBulkImportSession(String sessionName)
+    }
+
+    @Override
+    public TDSavedQuery updateSavedQuery(String name, TDSavedQueryBuilder request)
```
This method has a race condition where two concurrent requests might end up losing some saved query properties, even though they each intend to update separate properties.

Also it's a little inconsistent that it takes a builder instead of a request.

xerial replied:
@danielnorberg It seems there is no api that updates only a subset of job schedule properties. Or our internal documentation is wrong (I'll send a link to the doc later).

xerial replied:
Maybe we should create TDSavedQueryUpdateRequest class to make it consistent, but due to the same reason in the above, we need to merge the existing saved query data and its diff. 

No concurrent control will be available in this API design. We should note this limitation (or fix TD-API side)

xerial replied:
Added a comment for the race condition: 
https://github.com/treasure-data/td-client-java/pull/39/files#diff-f81246bf7411b6e6d93cfbfda2b0c7d2R681

danielnorberg commented on src/main/java/com/treasuredata/client/TDClientApi.java:
```diff
@@ -209,6 +216,22 @@
      */
     String startSavedQuery(String name, Date scheduledTime);
 
+    List<TDSavedQuery> listSavedQueries();
```
This interface is missing a `getSavedQuery(String name)` method, but the td rest api doesn't seem to have a method for fetching a single saved query. We should add one.

xerial replied:
We should do it :)

xerial commented:
@danielnorberg 
Could you review it again? 

I have applied your comments. Major differences are:
 - Added TDSavedQueryUpdateRequest 
 - update method becomes only TDClientApi.updateSavedQuery(name, TDSavedQueryUpdateRequest).
 - Added DSavedQuery.new(UpdateRequest)Builder

danielnorberg commented on src/main/java/com/treasuredata/client/TDClient.java:
```diff
@@ -660,24 +660,41 @@ protected String toJson(Object any)
         List<TDSavedQuery> savedQueries = listSavedQueries();
         for (TDSavedQuery q : savedQueries) {
             if (q.getName().equals(name)) {
+                /**
```
I'm still not sure it's great to have this method in a public library until we can support it properly. Even if we don't provide this method now, users (and we) can implement it themselves and thus be aware of the race condition pitfall.

I'll let you decide.

danielnorberg commented on src/main/java/com/treasuredata/client/TDClient.java:
```diff
@@ -627,14 +637,94 @@ public void deleteBulkImportSession(String sessionName)
+    }
+
+    @Override
+    public TDSavedQuery updateSavedQuery(String name, TDSavedQueryUpdateRequest request)
```
Looking at the api code, it actually seems like it does support partial update, so I believe we can implement this method in a safe manner.

danielnorberg replied:
Seems it's even possible to update the name.

xerial replied:
TDSavedQueryUpdateRequest can have a new query name.

xerial commented:
@danielnorberg Supported the partial updates of saved queries.

danielnorberg commented:
LGTM 👍

xerial commented:
Thanks for checking!
